tools/buildgen/extract_metadata_from_bazel_xml.py

```diff
@@ -937,31 +975,86 @@ def _generate_build_extra_metadata_for_tests(tests, bazel_rules):
-
-all_targets_dict = _generate_build_metadata(all_metadata, bazel_rules)
+# Step 2: Extract the known bazel cc_test tests. While most tests
+# will be buildable with other build systems just fine, some of these tests
+# would be too difficult to build and run with other build systems,
+# so we simply the ones we don't want.
+tests = _exclude_unwanted_cc_tests(_extract_cc_tests(bazel_rules))
+
+# Step 3: Generate the "extra metadata" for all our build targets.
+# While the bazel rules give us most of the information we need,
+# the legacy "build.yaml" format requires some additional fields that
+# we cannot get just from bazel alone (we call that "extra metadata").
+# In this step, we basically analyze the build metadata we have from bazel
+# and use heuristics to determine (and sometimes guess) the right
+# extra metadata to use for each target.
+#
+# - For some targets (such as the public libraries, helper libraries
+#   and executables) determining the right extra metadata is hard to do
+#   automatically. For these targets, the extra metadata is supplied "manually"
+#   in form of the _BUILD_EXTRA_METADATA dictionary. That allows us to match
+#   the semantics of the legacy "build.yaml" as closely as possible.
+#
+# - For test binaries, it is possible to generate the "extra metadata" mostly
+#   automatically using a rule-based heuristic approach because most tests
+#   look and behave alike from the build's perspective.
+#
+# TODO(jtattermusch): Of course neither "_BUILD_EXTRA_METADATA" or
+# the heuristic approach used for tests are ideal and they cannot be made
+# to cover all possible situations (and are tailored to work with the way
+# the grpc build currently works), but the idea was to start with something
+# reasonably simple that matches the "build.yaml"-like semantics as closely
+# as possible (to avoid changing too many things at once) and gradually get
+# rid of the legacy "build.yaml"-specific fields one by one. Once that is done,
+# only very little "extra metadata" would be needed and/or it would be trivial
+# to generate it automatically.
+all_extra_metadata = {}
+all_extra_metadata.update(_BUILD_EXTRA_METADATA)
+all_extra_metadata.update(
+    _generate_build_extra_metadata_for_tests(tests, bazel_rules))
+
+# Step 4: Generate the final metadata for all the targets.
+# This is done by combining the bazel build metadata and the "extra metadata"
+# we obtained in the previous step.
+# In this step, we also perform some interesting massaging of the target metadata
+# to end up with a result that is as similar to the legacy build.yaml data
+# as possible.
+# - Some targets get renamed (to match the legacy build.yaml target names)
+# - Some intermediate libraries get elided ("expanded") to better match the set
+#   of targets provided by the legacy build.yaml build
+all_targets_dict = _generate_build_metadata(all_extra_metadata, bazel_rules)
+
+# Step 5: convert the dictionary with all the targets to a dict that has
```

Comment:
Can you show an example of what `build_yaml_like` looks like after this step?

Reply:
done